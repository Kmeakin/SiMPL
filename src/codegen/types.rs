use crate::types::ty::Type;
use inkwell::{
    context::Context,
    types::{BasicType, BasicTypeEnum, FunctionType},
    AddressSpace,
};

impl Type {
    pub fn llvm_type<'ctx>(&self, ctx: &'ctx Context) -> BasicTypeEnum<'ctx> {
        match self {
            Self::Bool => ctx.bool_type().into(),
            Self::Int => ctx.i32_type().into(),
            Self::Float => ctx.f64_type().into(),
            Self::Fn(arg, ret) => ret
                .llvm_type(ctx)
                .fn_type(&[arg.llvm_type(ctx)], false)
                .ptr_type(AddressSpace::Global)
                .into(),
            Self::Var(_) => panic!("Cannot instantiate type {}", self),
        }
    }

    pub fn llvm_fn_type<'ctx>(&self, ctx: &'ctx Context) -> Option<FunctionType<'ctx>> {
        match self {
            Self::Fn(arg, ret) => ret
                .llvm_type(ctx)
                .fn_type(&[arg.llvm_type(ctx)], false)
                .into(),
            _ => None,
        }
    }
}
