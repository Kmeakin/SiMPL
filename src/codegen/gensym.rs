use crate::util::counter::{Counter, FromId};

#[derive(Debug, Copy, Clone)]
pub struct Gensym {
    counter: Counter<u32>,
    prefix: &'static str,
}

impl Gensym {
    pub fn new(prefix: &'static str) -> Self {
        Self {
            counter: Counter::new(),
            prefix,
        }
    }

    pub fn current(&self) -> String {
        format!("{}{}", self.prefix, self.current())
    }

    pub fn next(&mut self) -> String {
        let x = self.next();
        format!("{}{}", self.prefix, x)
    }
}
