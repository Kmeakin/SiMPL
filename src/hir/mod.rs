pub use crate::types::ty::Type;
use crate::{syntax::ast, types::ty::TypeVarGen};
use derive_more::Display;
pub use simple_symbol::Symbol;
use std::str::FromStr;

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Lit {
        ty: Type,
        val: Lit,
    },
    Var {
        ty: Type,
        name: Symbol,
    },
    If {
        ty: Type,
        test: Box<Self>,
        then: Box<Self>,
        els: Box<Self>,
    },
    Let {
        ty: Type,
        binding: LetBinding,
        body: Box<Self>,
    },
    Letrec {
        ty: Type,
        bindings: Vec<LetBinding>,
        body: Box<Self>,
    },
    Lambda {
        ty: Type,
        param: Param,
        body: Box<Self>,
    },
    App {
        ty: Type,
        func: Box<Self>,
        arg: Box<Self>,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub struct Param {
    pub ty: Type,
    pub name: Symbol,
    pub ann: Option<Type>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct LetBinding {
    pub ty: Type,
    pub name: Symbol,
    pub ann: Option<Type>,
    pub val: Box<Expr>,
}

#[derive(Debug, Copy, Clone, PartialEq, Display)]
pub enum Lit {
    #[display(fmt = "{}", _0)]
    Bool(bool),
    #[display(fmt = "{}", _0)]
    Int(i64),
    #[display(fmt = "{}", _0)]
    Float(f64),
}

impl From<bool> for Lit {
    fn from(other: bool) -> Self {
        Self::Bool(other)
    }
}

impl From<i64> for Lit {
    fn from(other: i64) -> Self {
        Self::Int(other)
    }
}

impl From<f64> for Lit {
    fn from(other: f64) -> Self {
        Self::Float(other)
    }
}

impl From<ast::Lit> for Lit {
    fn from(other: ast::Lit) -> Self {
        match other {
            ast::Lit::Int(x) => Self::Int(x),
            ast::Lit::Float(x) => Self::Float(x),
            ast::Lit::Bool(x) => Self::Bool(x),
        }
    }
}

impl FromStr for Expr {
    type Err = String;
    fn from_str(src: &str) -> Result<Self, Self::Err> {
        // TODO: return a trait object instead of unwrapping
        let ast = crate::syntax::parse(src).unwrap();
        Ok(Self::from_ast(ast))
    }
}

impl Expr {
    /// `ast::Expr` -> `hir::Expr`
    /// Expands nested let/lambda, and attatches fresh type variables to every
    /// expression/binder
    pub fn from_ast(ast: ast::Expr) -> Self {
        let mut gen = TypeVarGen::new();
        Self::from_ast_inner(ast, &mut gen)
    }

    fn from_ast_inner(ast: ast::Expr, gen: &mut TypeVarGen) -> Self {
        match ast {
            ast::Expr::Lit { val } => Self::Lit {
                val: val.into(),
                ty: gen.next(),
            },
            ast::Expr::Var { name } => Self::Var {
                name,
                ty: gen.next(),
            },
            ast::Expr::Binop { .. } => todo!(),
            ast::Expr::If { test, then, els } => Self::If {
                ty: gen.next(),
                test: box Self::from_ast_inner(*test, gen),
                then: box Self::from_ast_inner(*then, gen),
                els: box Self::from_ast_inner(*els, gen),
            },
            ast::Expr::Let { bindings, body } => {
                let (binding, body) = expand_let(&bindings, *body);
                Self::Let {
                    ty: gen.next(),
                    binding: LetBinding {
                        ty: gen.next(),
                        name: binding.name,
                        ann: binding.ann,
                        val: box Self::from_ast_inner(*binding.val, gen),
                    },
                    body: box Self::from_ast_inner(body, gen),
                }
            }
            ast::Expr::Letrec { bindings, body } => Self::Letrec {
                ty: gen.next(),
                bindings: bindings
                    .into_iter()
                    .map(|binding| LetBinding {
                        ty: gen.next(),
                        name: binding.name,
                        ann: binding.ann,
                        val: box Self::from_ast_inner(*binding.val, gen),
                    })
                    .collect(),
                body: box Self::from_ast_inner(*body, gen),
            },
            ast::Expr::Lambda { params, body } => {
                let (param, body) = expand_lambda(&params, *body);
                Self::Lambda {
                    ty: gen.next(),
                    param: Param {
                        name: param.name,
                        ann: param.ann,
                        ty: gen.next(),
                    },
                    body: box Self::from_ast_inner(body, gen),
                }
            }
            ast::Expr::App { func, arg } => Self::App {
                ty: gen.next(),
                func: box Self::from_ast_inner(*func, gen),
                arg: box Self::from_ast_inner(*arg, gen),
            },
        }
    }

    pub fn ty(&self) -> Type {
        match self {
            Self::Lit { ty, .. }
            | Self::Var { ty, .. }
            | Self::If { ty, .. }
            | Self::Let { ty, .. }
            | Self::Letrec { ty, .. }
            | Self::Lambda { ty, .. }
            | Self::App { ty, .. } => ty.clone(),
        }
    }
}

fn expand_lambda(params: &[ast::Param], body: ast::Expr) -> (ast::Param, ast::Expr) {
    assert!(!params.is_empty());
    if params.len() == 1 {
        (params[0].clone(), body)
    } else {
        let param = &params[0];
        let (rest_param, rest_body) = expand_lambda(&params[1..], body);
        (
            param.clone(),
            ast::Expr::Lambda {
                params: vec![rest_param],
                body: box rest_body,
            },
        )
    }
}

fn expand_let(bindings: &[ast::LetBinding], body: ast::Expr) -> (ast::LetBinding, ast::Expr) {
    assert!(!bindings.is_empty());
    if bindings.len() == 1 {
        (bindings[0].clone(), body)
    } else {
        let binding = &bindings[0];
        let (rest_bindings, rest_body) = expand_let(&bindings[1..], body);
        (
            binding.clone(),
            ast::Expr::Let {
                bindings: vec![rest_bindings],
                body: box rest_body,
            },
        )
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::syntax::ast::{Expr, Lit};
    use simple_symbol::intern;

    #[test]
    fn test_expand_lambda() {
        let params = vec![
            ast::Param {
                name: intern("x"),
                ann: None,
            },
            ast::Param {
                name: intern("y"),
                ann: None,
            },
        ];
        let body = Expr::Lit { val: Lit::Int(0) };

        assert_eq!(
            expand_lambda(&params, body),
            (
                ast::Param {
                    name: intern("x"),
                    ann: None
                },
                Expr::Lambda {
                    params: vec![ast::Param {
                        name: intern("y"),
                        ann: None
                    }],
                    body: box Expr::Lit { val: Lit::Int(0) }
                }
            ),
        )
    }

    #[test]
    fn test_expand_let() {
        let bindings = vec![
            ast::LetBinding {
                name: intern("x"),
                val: box Expr::Lit { val: Lit::Int(1) },
                ann: None,
            },
            ast::LetBinding {
                name: intern("y"),
                val: box Expr::Lit { val: Lit::Int(2) },
                ann: None,
            },
        ];

        let body = Expr::Lit { val: Lit::Int(0) };

        assert_eq!(
            expand_let(&bindings, body),
            (
                ast::LetBinding {
                    name: intern("x"),
                    val: box Expr::Lit { val: Lit::Int(1) },
                    ann: None,
                },
                Expr::Let {
                    bindings: vec![ast::LetBinding {
                        name: intern("y"),
                        val: box Expr::Lit { val: Lit::Int(2) },
                        ann: None,
                    }],
                    body: box Expr::Lit { val: Lit::Int(0) }
                }
            ),
        )
    }
}
