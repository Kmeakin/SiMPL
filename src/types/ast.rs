pub use crate::syntax::ast::{Ident, Lit};
use crate::{
    syntax::ast::Expr,
    types::ty::{Type, TypeEnv},
    util::counter::{Counter, FromId},
};
use std::str::FromStr;

#[derive(Debug, Clone, PartialEq)]
pub enum TypedExpr {
    Lit {
        ty: Type,
        val: Lit,
    },
    Var {
        ty: Type,
        name: String,
    },
    If {
        ty: Type,
        test: Box<Self>,
        then_branch: Box<Self>,
        else_branch: Box<Self>,
    },
    Let {
        ty: Type,
        binding: LetBinding,
        body: Box<Self>,
    },
    Letrec {
        ty: Type,
        bindings: Vec<LetBinding>,
        body: Box<Self>,
    },
    Lambda {
        ty: Type,
        param: Param,
        body: Box<Self>,
    },
    App {
        ty: Type,
        func: Box<Self>,
        arg: Box<Self>,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub struct Param {
    pub ty: Type,
    pub name: Ident,
}

#[derive(Debug, Clone, PartialEq)]
pub struct LetBinding {
    pub ty: Type,
    pub name: Ident,
    pub val: Box<TypedExpr>,
}

pub type TypeVarGen = Counter<Type>;
impl FromId for Type {
    fn from_id(id: u32) -> Self {
        Self::Var(id)
    }
}

impl FromStr for TypedExpr {
    type Err = String;
    fn from_str(src: &str) -> Result<Self, Self::Err> {
        // TODO: return a trait object instead of unwrapping
        let ast = crate::syntax::parse(src).unwrap();
        Self::from_ast(ast)
    }
}

impl TypedExpr {
    pub fn from_ast(ast: Expr) -> Result<Self, String> {
        let mut gen = TypeVarGen::new();
        let tenv = TypeEnv::default();
        Self::from_ast_inner(ast, &tenv, &mut gen)
    }

    fn from_ast_inner(ast: Expr, tenv: &TypeEnv, gen: &mut TypeVarGen) -> Result<Self, String> {
        let expr = match ast {
            Expr::Lit { val } => Self::Lit {
                val,
                ty: gen.next(),
            },
            Expr::Var { name } => match tenv.get(&name) {
                None => return Err(format!("Unbound variable: {}", name)),
                Some(ty) => Self::Var {
                    ty: ty.clone(),
                    name,
                },
            },
            Expr::If {
                test,
                then_branch,
                else_branch,
            } => Self::If {
                ty: gen.next(),
                test: box Self::from_ast_inner(*test, tenv, gen)?,
                then_branch: box Self::from_ast_inner(*then_branch, tenv, gen)?,
                else_branch: box Self::from_ast_inner(*else_branch, tenv, gen)?,
            },
            Expr::Let { bindings, body } => {
                let (bindings, body) = expand_let(&bindings, *body);

                assert!(bindings.len() == 1);
                let ty = gen.next();

                let (binding_name, binding_val) = &bindings[0];
                let binding_ty = gen.next();
                let mut extended_tenv = tenv.clone();
                extended_tenv.insert(binding_name.into(), binding_ty.clone());

                Self::Let {
                    ty,
                    binding: LetBinding {
                        ty: binding_ty,
                        name: binding_name.clone(),
                        val: box Self::from_ast_inner(binding_val.clone(), tenv, gen)?,
                    },
                    body: box Self::from_ast_inner(body, &extended_tenv, gen)?,
                }
            }
            Expr::Letrec { bindings, body } => {
                let ty = gen.next();
                let mut extended_tenv = tenv.clone();

                let mut tvars = vec![];
                for (name, _) in &bindings {
                    let tvar = gen.next();
                    extended_tenv.insert(name.clone(), tvar.clone());
                    tvars.push(tvar);
                }

                let mut new_bindings = vec![];

                for ((name, val), tvar) in bindings.into_iter().zip(tvars) {
                    new_bindings.push(LetBinding {
                        name: name.clone(),
                        ty: tvar,
                        val: box Self::from_ast_inner(val.clone(), &extended_tenv, gen)?,
                    });
                }

                Self::Letrec {
                    ty,
                    bindings: new_bindings,
                    body: box Self::from_ast_inner(*body, &extended_tenv, gen)?,
                }
            }
            Expr::Lambda { params, body } => {
                let (params, body) = expand_lambda(&params, *body);

                assert!(params.len() == 1);
                let ty = gen.next();

                let param_name = &params[0];
                let param_ty = gen.next();
                let mut extended_tenv = tenv.clone();
                extended_tenv.insert(param_name.clone(), param_ty.clone());
                Self::Lambda {
                    ty,
                    param: Param {
                        ty: param_ty,
                        name: param_name.clone(),
                    },
                    body: box Self::from_ast_inner(body, &extended_tenv, gen)?,
                }
            }
            Expr::App { func, arg } => Self::App {
                ty: gen.next(),
                func: box Self::from_ast_inner(*func, tenv, gen)?,
                arg: box Self::from_ast_inner(*arg, tenv, gen)?,
            },
        };
        Ok(expr)
    }

    pub fn ty(&self) -> Type {
        match self {
            Self::Lit { ty, .. }
            | Self::Var { ty, .. }
            | Self::If { ty, .. }
            | Self::Let { ty, .. }
            | Self::Letrec { ty, .. }
            | Self::Lambda { ty, .. }
            | Self::App { ty, .. } => ty.clone(),
        }
    }
}

/// Expand an `ast::Expr::Let` with many bound variables into a nested
/// sequences of lets each binding a single variable
/// eg `let x = 1, y = 2 in add x y` expands to
///    `let x = 1 in let y = 2 in add x y`
fn expand_let(bindings: &[(Ident, Expr)], body: Expr) -> (Vec<(Ident, Expr)>, Expr) {
    assert!(!bindings.is_empty());
    if bindings.len() == 1 {
        (bindings.to_vec(), body)
    } else {
        let binding = &bindings[0];

        let (rest_bindings, rest_body) = expand_let(&bindings[1..], body);

        (
            vec![binding.clone()],
            Expr::Let {
                bindings: rest_bindings,
                body: box rest_body,
            },
        )
    }
}

/// Expand an `ast::Expr::Lambda` with many parameters into a nested
/// sequences of lambdas each binding a single parameter
/// eg `\x, y -> add x y` expands to
///    `\x -> \y -> add x y`
fn expand_lambda(params: &[Ident], body: Expr) -> (Vec<Ident>, Expr) {
    assert!(!params.is_empty());
    if params.len() == 1 {
        (params.to_vec(), body)
    } else {
        let param = &params[0];

        let (rest_params, rest_body) = expand_lambda(&params[1..], body);

        (
            vec![param.clone()],
            Expr::Lambda {
                params: rest_params,
                body: box rest_body,
            },
        )
    }
}

#[test]
fn test_expand_let() {
    let bindings = vec![
        ("x".into(), Expr::Lit { val: Lit::Int(1) }),
        ("y".into(), Expr::Lit { val: Lit::Int(2) }),
    ];

    let body = Expr::Lit { val: Lit::Int(0) };

    assert_eq!(
        expand_let(&bindings, body),
        (
            vec![("x".into(), Expr::Lit { val: Lit::Int(1) }),],
            Expr::Let {
                bindings: vec![("y".into(), Expr::Lit { val: Lit::Int(2) })],
                body: box Expr::Lit { val: Lit::Int(0) }
            }
        ),
    )
}

#[test]
fn test_expand_lambda() {
    let params = vec!["x".into(), "y".into()];
    let body = Expr::Lit { val: Lit::Int(0) };

    assert_eq!(
        expand_lambda(&params, body),
        (
            vec!["x".into()],
            Expr::Lambda {
                params: vec!["y".into()],
                body: box Expr::Lit { val: Lit::Int(0) }
            }
        ),
    )
}
