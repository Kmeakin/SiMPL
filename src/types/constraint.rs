use crate::{
    hir::{Binop, Expr},
    types::ty::{Type, TypeEnv},
};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Constraint(pub(crate) Type, pub(crate) Type);
pub type Constraints = Vec<Constraint>;

/// Collect constraints, and check for unbound variables
pub fn collect(expr: Expr) -> Constraints {
    let tenv = TypeEnv::default();
    collect_inner(expr, &tenv)
}

fn collect_inner(expr: Expr, tenv: &TypeEnv) -> Constraints {
    match expr {
        Expr::Lit { ty, val } => vec![Constraint(ty, val.ty())],
        Expr::Var { ty, name } => match tenv.get(name) {
            Some(ty2) => vec![Constraint(ty, ty2.clone())],
            None => panic!("Unbound variable: {}", name),
        },
        Expr::Binop { ty, lhs, rhs, op } => {
            use Binop::*;
            use Type::*;

            let (lhs_ty, rhs_ty, out_ty) = match op {
                IntAdd | IntSub | IntMul | IntDiv => (Int, Int, Int),
                IntLt | IntLeq | IntGt | IntGeq => (Int, Int, Bool),

                FloatAdd | FloatSub | FloatMul | FloatDiv => (Float, Float, Float),
                FloatLt | FloatLeq | FloatGt | FloatGeq => (Float, Float, Bool),

                Eq | Neq => (rhs.ty(), lhs.ty(), Bool),
            };

            let mut cons = vec![
                Constraint(lhs.ty(), lhs_ty),
                Constraint(rhs.ty(), rhs_ty),
                Constraint(ty, out_ty),
            ];

            cons.extend(collect_inner(*lhs, tenv));
            cons.extend(collect_inner(*rhs, tenv));
            cons
        }
        Expr::If {
            ty,
            test,
            then,
            els,
        } => {
            let mut cons = vec![
                Constraint(test.ty(), Type::Bool),
                Constraint(then.ty(), ty.clone()),
                Constraint(els.ty(), ty),
            ];
            cons.extend(collect_inner(*test, tenv));
            cons.extend(collect_inner(*then, tenv));
            cons.extend(collect_inner(*els, tenv));
            cons
        }

        Expr::Let { ty, binding, body } => {
            let mut ext_tenv = tenv.clone();
            ext_tenv.insert(binding.name, binding.ty.clone());

            let mut cons = vec![
                Constraint(ty, body.ty()),
                Constraint(binding.ty, binding.val.ty()),
            ];

            if let Some(ty) = binding.ann {
                cons.push(Constraint(ty, binding.val.ty()));
            }

            cons.extend(collect_inner(*binding.val, tenv));
            cons.extend(collect_inner(*body, &ext_tenv));
            cons
        }
        Expr::Letrec { ty, bindings, body } => {
            assert!(!bindings.is_empty());

            let mut ext_tenv = tenv.clone();
            let mut cons = vec![Constraint(ty, body.ty())];

            for binding in &bindings {
                cons.push(Constraint(binding.ty.clone(), binding.val.ty()));
                ext_tenv.insert(binding.name, binding.ty.clone())
            }

            for binding in bindings {
                cons.extend(collect_inner(*binding.val.clone(), &ext_tenv));
            }

            cons.extend(collect_inner(*body, &ext_tenv));
            cons
        }
        Expr::Lambda { ty, param, body } => {
            let mut ext_tenv = tenv.clone();
            ext_tenv.insert(param.name, param.ty.clone());

            let mut cons = vec![Constraint(
                ty,
                Type::Fn(box param.ty.clone(), box body.ty()),
            )];

            if let Some(ty) = param.ann {
                cons.push(Constraint(ty, param.ty))
            }

            cons.extend(collect_inner(*body, &ext_tenv));
            cons
        }
        Expr::App { ty, func, arg } => {
            let mut cons = vec![Constraint(func.ty(), Type::Fn(box arg.ty(), box ty))];
            cons.extend(collect_inner(*func, tenv));
            cons.extend(collect_inner(*arg, tenv));
            cons
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::{
        hir::{LetBinding, Lit, Param},
        ty,
        types::ty::TypeVarGen,
    };
    use simple_symbol::intern;

    #[test]
    fn constrain_int() {
        let mut gen = TypeVarGen::new();

        let t1 = gen.next();
        let expr = Expr::Lit {
            ty: t1.clone(),
            val: Lit::Int(1),
        };
        assert_eq!(collect(expr), vec![Constraint(t1, Type::Int)]);
    }

    #[test]
    fn constrain_bool() {
        let mut gen = TypeVarGen::new();

        let t1 = gen.next();
        let expr = Expr::Lit {
            ty: t1.clone(),
            val: Lit::Bool(true),
        };
        assert_eq!(collect(expr), vec![Constraint(t1, Type::Bool)]);
    }

    #[test]
    fn constrain_float() {
        let mut gen = TypeVarGen::new();

        let t1 = gen.next();
        let expr = Expr::Lit {
            ty: t1.clone(),
            val: Lit::Float(1.23),
        };
        assert_eq!(collect(expr), vec![Constraint(t1, Type::Float)]);
    }

    #[test]
    fn constrain_lambda() {
        let mut gen = TypeVarGen::new();

        let t0 = gen.next();
        let t1 = gen.next();
        let t2 = gen.next();

        let expr = Expr::Lambda {
            ty: t0.clone(),
            param: Param {
                ty: t1.clone(),
                name: intern("param").into(),
                ann: None,
            },
            body: box Expr::Var {
                ty: t2.clone(),
                name: intern("param").into(),
            },
        };
        assert_eq!(
            collect(expr),
            vec![Constraint(t0, ty![{1} => {2}]), Constraint(t2, t1)],
        );
    }

    #[test]
    fn constrain_var() {
        let mut gen = TypeVarGen::new();

        let t0 = gen.next();

        let expr = Expr::Var {
            ty: t0.clone(),
            name: intern("not").into(),
        };

        assert_eq!(collect(expr), vec![Constraint(t0, ty![Bool => Bool])]);
    }

    #[test]
    fn constrain_app() {
        let mut gen = TypeVarGen::new();

        let t0 = gen.next();
        let t1 = gen.next();
        let t2 = gen.next();

        let expr = Expr::App {
            ty: t0.clone(),
            func: box Expr::Var {
                ty: t1.clone(),
                name: intern("add").into(),
            },
            arg: box Expr::Lit {
                ty: t2.clone(),
                val: Lit::Int(0),
            },
        };

        assert_eq!(
            collect(expr),
            vec![
                Constraint(t1.clone(), ty![{2} => {0}]),
                Constraint(t1.clone(), ty![Int => Int => Int]),
                Constraint(t2.clone(), ty![Int])
            ]
        );
    }

    #[test]
    fn constrain_let() {
        let mut gen = TypeVarGen::new();

        let t0 = gen.next();
        let t1 = gen.next();
        let t2 = gen.next();
        let t3 = gen.next();

        let expr = Expr::Let {
            ty: t0.clone(),
            binding: LetBinding {
                ty: t1.clone(),
                name: intern("name").into(),
                val: box Expr::Lit {
                    ty: t2.clone(),
                    val: Lit::Bool(false),
                },
                ann: None,
            },
            body: box Expr::Var {
                ty: t3.clone(),
                name: intern("name").into(),
            },
        };

        assert_eq!(
            collect(expr),
            vec![
                Constraint(t0, t3.clone()),
                Constraint(t1.clone(), t2.clone()),
                Constraint(t2, Type::Bool),
                Constraint(t3, t1),
            ]
        );
    }

    #[test]
    fn constrain_identity() {
        let mut gen = TypeVarGen::new();

        let t0 = gen.next();
        let t1 = gen.next();
        let t2 = gen.next();

        let expr = Expr::Lambda {
            ty: t0.clone(),
            param: Param {
                ty: t1.clone(),
                name: intern("a").into(),
                ann: None,
            },
            body: box Expr::Var {
                ty: t2.clone(),
                name: intern("a").into(),
            },
        };

        assert_eq!(
            collect(expr),
            vec![Constraint(t0, ty![{1} => {2}]), Constraint(t2, t1)]
        );
    }

    #[test]
    fn constrain_const() {
        let mut gen = TypeVarGen::new();
        let t0 = gen.next();
        let t1 = gen.next();
        let t2 = gen.next();
        let t3 = gen.next();
        let t4 = gen.next();

        let expr = Expr::Lambda {
            ty: t0.clone(),
            param: Param {
                ty: t1.clone(),
                name: intern("a").into(),
                ann: None,
            },
            body: box Expr::Lambda {
                ty: t2.clone(),
                param: Param {
                    ty: t3.clone(),
                    name: intern("b").into(),
                    ann: None,
                },
                body: box Expr::Var {
                    ty: t4.clone(),
                    name: intern("a").into(),
                },
            },
        };

        assert_eq!(
            collect(expr),
            vec![
                Constraint(t0, ty![{1} => {2}]),
                Constraint(t2, ty![{3} => {4}]),
                Constraint(t4, t1)
            ]
        )
    }

    #[test]
    fn constrain_compose() {
        let mut gen = TypeVarGen::new();
        let t0 = gen.next();
        let t1 = gen.next();
        let t2 = gen.next();
        let t3 = gen.next();
        let t4 = gen.next();
        let t5 = gen.next();
        let t6 = gen.next();
        let t7 = gen.next();
        let t8 = gen.next();
        let t9 = gen.next();
        let t10 = gen.next();

        let expr = Expr::Lambda {
            ty: t0.clone(),
            param: Param {
                ty: t1.clone(),
                name: intern("f").into(),
                ann: None,
            },
            body: box Expr::Lambda {
                ty: t2.clone(),
                param: Param {
                    ty: t3.clone(),
                    name: intern("g").into(),
                    ann: None,
                },
                body: box Expr::Lambda {
                    ty: t4.clone(),
                    param: Param {
                        ty: t5.clone(),
                        name: intern("x").into(),
                        ann: None,
                    },
                    body: box Expr::App {
                        ty: t6.clone(),
                        func: box Expr::Var {
                            ty: t7.clone(),
                            name: intern("f").into(),
                        },
                        arg: box Expr::App {
                            ty: t8.clone(),
                            func: box Expr::Var {
                                ty: t9.clone(),
                                name: intern("g").into(),
                            },
                            arg: box Expr::Var {
                                ty: t10.clone(),
                                name: intern("x").into(),
                            },
                        },
                    },
                },
            },
        };

        assert_eq!(
            collect(expr),
            vec![
                Constraint(t0.clone(), Type::Fn(box t1.clone(), box t2.clone())),
                Constraint(t2.clone(), Type::Fn(box t3.clone(), box t4.clone())),
                Constraint(t4.clone(), Type::Fn(box t5.clone(), box t6.clone())),
                Constraint(t7.clone(), Type::Fn(box t8.clone(), box t6.clone())),
                Constraint(t7.clone(), t1.clone()),
                Constraint(t9.clone(), Type::Fn(box t10.clone(), box t8.clone())),
                Constraint(t9.clone(), t3.clone()),
                Constraint(t10.clone(), t5.clone()),
            ]
        );
    }
}
