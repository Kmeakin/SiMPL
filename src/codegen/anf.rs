//! `TypedExpr` -> `TypedExpr` pass
//! Assumes the rename pass has already run
//!
//! Input language:
//! e := x
//!    | n
//!    | if e then e else e
//!    | let x = e in e
//!    | letrec (x = e)+ in e
//!    | \x -> e
//!    | e e
//!
//! x := identifer
//!
//! n := int
//!    | float
//!    | bool
//!
//! Output language:
//! e := ae
//!    | ae ae
//!    | if ae e e
//!    | let x = e in e
//!
//! ae := \x -> e
//!     | x
//!     | n

use super::gensym::Gensym;
use crate::hir::{Expr, LetBinding};
use lazy_static::lazy_static;
use std::sync::Mutex;

lazy_static! {
    static ref GENSYM: Mutex<Gensym> = Mutex::new(Gensym::new("$"));
}

impl Expr {
    fn is_anf(&self) -> bool {
        self.is_e()
    }

    fn is_e(&self) -> bool {
        match self {
            Self::If {
                test,
                then_branch,
                else_branch,
                ..
            } => test.is_ae() && then_branch.is_ae() && else_branch.is_ae(),
            Self::Let { binding, body, .. } => binding.val.is_e() && body.is_e(),
            Self::Letrec { .. } => todo!(),
            Self::App { func, arg, .. } => func.is_ae() && arg.is_ae(),
            _ => self.is_ae(),
        }
    }

    fn is_ae(&self) -> bool {
        match self {
            Self::Lambda { body, .. } => body.is_e(),
            Self::Lit { .. } | Self::Var { .. } => true,
            _ => false,
        }
    }
}

pub fn normalize_expr(expr: Expr) -> Expr {
    GENSYM.lock().unwrap().reset();
    normalize_expr_inner(expr)
}

fn normalize_expr_inner(expr: Expr) -> Expr {
    normalize(expr, box |x| x)
}

fn normalize(expr: Expr, k: Box<dyn FnOnce(Expr) -> Expr>) -> Expr {
    match expr {
        Expr::If {
            ty,
            test,
            then_branch,
            else_branch,
        } => normalize_name(*test, box |t| {
            k(Expr::If {
                ty,
                test: box t,
                then_branch: box normalize_expr_inner(*then_branch),
                else_branch: box normalize_expr_inner(*else_branch),
            })
        }),

        Expr::Let { ty, binding, body } => normalize(*binding.clone().val, box |n1| Expr::Let {
            ty,
            binding: LetBinding {
                val: box n1,
                ..binding
            },
            body: box normalize(*body, k),
        }),

        Expr::Letrec { .. } => todo!(),

        Expr::Lambda { ty, param, body } => k(Expr::Lambda {
            ty,
            param,
            body: box normalize_expr_inner(*body),
        }),

        Expr::App { ty, func, arg } => normalize_name(*func, box |t| {
            normalize_name(*arg, box |t2| {
                k(Expr::App {
                    ty,
                    func: box t,
                    arg: box t2,
                })
            })
        }),

        Expr::Lit { .. } | Expr::Var { .. } => k(expr),
    }
}

fn normalize_name(expr: Expr, k: Box<dyn FnOnce(Expr) -> Expr>) -> Expr {
    normalize(expr, box |n| match n {
        Expr::Lit { .. } | Expr::Var { .. } => k(n),
        _ => {
            let name = GENSYM.lock().unwrap().next();
            let ty = n.ty();
            let t = Expr::Var {
                ty: ty.clone(),
                name,
            };
            Expr::Let {
                ty: ty.clone(),
                binding: LetBinding {
                    ty,
                    name,
                    val: box n,
                },
                body: box k(t),
            }
        }
    })
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::types::{self, parse_and_type};
    use insta::assert_snapshot;

    #[track_caller]
    fn test_normalize(src: &str) {
        let expr = parse_and_type(src);
        let norm = normalize_expr(expr);
        assert!(norm.is_anf());
        assert_snapshot!(norm.pretty());
    }

    #[test]
    fn normalize_lit() {
        test_normalize("123");
    }

    #[test]
    fn normalize_if() {
        test_normalize("if true then 1 else 0");
        test_normalize("if (if false then true else false) then 1 else 0");
    }

    // TODO: should immediate values be allowed in the value of a binding?
    #[test]
    fn normalize_let() {
        test_normalize("let abc = 1 in abc");
    }

    #[test]
    fn normalize_lambda() {
        test_normalize(r"\x -> x");
        test_normalize(r"\x -> 1");
    }

    #[test]
    fn normalize_app() {
        test_normalize(r"add (mul 2 4) (mul 2 8)");
    }
}
