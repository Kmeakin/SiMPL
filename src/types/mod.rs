use self::ty::Type;
use crate::hir::Expr;
use std::str::FromStr;

mod constraint;
mod subst;
pub mod ty;
mod unify;

mod pp;

#[cfg(test)]
mod test;

/// Infer the type of the expr
pub fn type_of(expr: &Expr) -> Type {
    let cons = constraint::collect(expr.clone());
    let subst = unify::unify(&cons);
    subst.apply_ty(&expr.ty())
}

/// Infer the type of the expr, and apply the resulting substitution to the
/// expression (so every expr has its inferred type attatched)
pub fn infer_and_apply(expr: &Expr) -> Expr {
    let cons = constraint::collect(expr.clone());
    let subst = unify::unify(&cons);
    expr.apply(&subst)
}

/// Convenience function. Parse source code, and give every expr its inferred
/// type
pub fn parse_and_type(src: &str) -> Expr {
    // TODO: return a &dyn impl Error instead of unwrapping
    let expr = Expr::from_str(src).unwrap();
    infer_and_apply(&expr)
}
