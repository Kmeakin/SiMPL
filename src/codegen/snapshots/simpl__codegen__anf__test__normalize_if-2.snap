---
source: src/codegen/anf.rs
expression: norm.pretty()
---
let gensym = if false
        then true
        else false in
    if gensym then 1 else 0
